internal/tripper/chain_test.go

```diff
@@ -0,0 +1,123 @@
+package tripper // import "github.com/pomerium/pomerium/internal/tripper"
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+)
+
+type mockTransport struct {
+	id string
+}
+
+func (t *mockTransport) RoundTrip(r *http.Request) (*http.Response, error) {
+	w := httptest.NewRecorder()
+
+	w.WriteString(t.id)
+	return w.Result(), nil
+}
+
+// mockMiddleware appends the id into the response body as
+// the call stack unwinds.
+//
+// If your chain is c1->c2->t, it should return 't,c2,c1'
+func mockMiddleware(id string) func(next http.RoundTripper) http.RoundTripper {
+	return func(next http.RoundTripper) http.RoundTripper {
+		return RoundTripperFunc(func(r *http.Request) (*http.Response, error) {
+
+			resp, _ := next.RoundTrip(r)
+
+			body, _ := ioutil.ReadAll(resp.Body)
+			mockResp := httptest.NewRecorder()
+			mockResp.Write(body)
+			mockResp.WriteString(fmt.Sprintf(",%s", id))
+			return mockResp.Result(), nil
+		})
+	}
+}
+
+func TestNew(t *testing.T) {
```

Comment:
The tripper chain functionality is largely copied from middleware, but I wanted to redo the tests from first principles so that (a) I understood the code well (b) could focus on the side effects rather than returned data structures.  

I think this approach captured all the tests nicely, but I'm happy to port over the middleware chain tests if you'd like to keep it consistent.

Reply:
I think that makes sense. I'm generally for a bit of duplication for the sake of clarity and I think that's true here.

Reply:
Do you want me to bring over the middleware tests or leave the tests as-is?  I'm okay either way.

Reply:
Also an option: update the middleware tests to use this style.

Reply:
Hmmm... If you think the middleware tests are relevant -- sure. If not, I don't think it's required. I'll defer to your judgement. 

The one thing that would be nice to add (sorry, I can't see coverage directly here as I'm traveling, so perhaps you already have that base covered) would be to have a mocked RoundTripper that also could test for that nil condition from before. .... maybe add a `roundTripError` type to `mockTransport`?

As for changing the existing middleware tests format, I'm a bit more cautious just because they are a complete carry over from [Alice](https://github.com/pomerium/pomerium/blob/master/3RD-PARTY#L90_ and have been pretty stable. Though, adding additional relevant tests makes sense.

Reply:
>have a mocked RoundTripper that also could test for that nil condition from before

Yep, should be covered.  I have a then() transport/tripper that errors. 
 https://github.com/travisgroth/pomerium/blob/ca46202d21561eab90c5a59113080a5c5190b6c3/internal/metrics/middleware_test.go#L128-L132  and https://github.com/travisgroth/pomerium/blob/ca46202d21561eab90c5a59113080a5c5190b6c3/internal/metrics/middleware_test.go#L215-L221

>As for changing the existing middleware tests format, I'm a bit more cautious just because they are a complete carry over from [Alice]

Ah, I did not realize.  I think they accomplish the same thing; just a different approach.  Take a look when you're back at a keyboard for a bit.  I'm happy to leave as is; we can always revisit.

Reply:
Ah! Figured that I just missed it.

Reply:
> I think they accomplish the same thing; just a different approach. Take a look when you're back at a keyboard for a bit. I'm happy to leave as is; we can always revisit.

Yeah, let's revisit in another issue.